File: attendance-server/app/core/permissions.py
"""
app/core/permissions.py

The permission catalogue — the exact same list as the frontend's
contracts/permissions.ts, and the source the seed script writes into the
database.

A permission is a gate written into the application, in two places: the screen
that hides a control, and the endpoint that refuses the request. One inserted at
runtime would be consulted by nothing — a checkbox that appears to lock
something down while locking down nothing, which is worse than no checkbox.

So this list lives in code. New permissions arrive with new features, because
the feature is what enforces them.
"""

from typing import NamedTuple


class PermissionDef(NamedTuple):
    code: str
    module: str
    label: str
    description: str


PERMISSIONS: list[PermissionDef] = [
    # --- Employees ---------------------------------------------------------
    PermissionDef("employees.view_all", "Employees", "View all",
                  "See the whole company directory, with contact details, department and login status."),
    PermissionDef("employees.view_own", "Employees", "View own",
                  "See only their own employee record. Everyone needs this to open their profile."),
    PermissionDef("employees.create", "Employees", "Create",
                  "Add a new person to the directory."),
    PermissionDef("employees.edit", "Employees", "Edit",
                  "Change anyone's details, or mark them inactive."),

    # --- Attendance --------------------------------------------------------
    PermissionDef("attendance.view_all", "Attendance", "View all",
                  "See every employee's check-in and check-out records and the exception list."),
    PermissionDef("attendance.view_own", "Attendance", "View own",
                  "See only their own attendance history."),
    PermissionDef("attendance.punch_manual", "Attendance", "Manual punch",
                  "Record a punch by hand when the kiosk missed someone. Entries stay marked as manual."),

    # --- Tasks -------------------------------------------------------------
    PermissionDef("tasks.view_all", "Tasks", "View all",
                  "See tasks assigned to everyone."),
    PermissionDef("tasks.view_own", "Tasks", "View own",
                  "See only tasks assigned to themselves."),
    PermissionDef("tasks.assign", "Tasks", "Assign",
                  "Create a task for somebody else."),
    PermissionDef("tasks.create_own", "Tasks", "Add own",
                  "Add a task to their own list."),
    PermissionDef("tasks.edit", "Tasks", "Edit",
                  "Change a task's title, description, due date or priority."),
    PermissionDef("tasks.complete_own", "Tasks", "Complete own",
                  "Tick off a task assigned to them."),

    # --- Administration ----------------------------------------------------
    PermissionDef("roles.manage", "Administration", "Manage roles",
                  "Create roles and change what every role can do. At least one role in use must keep it."),
    PermissionDef("users.manage", "Administration", "Manage users",
                  "Create login accounts, reset passwords, enable or disable access."),
    PermissionDef("users.change_role", "Administration", "Change role",
                  "Move someone to a different role, limited to roles no more powerful than your own."),
    # --- Leave -------------------------------------------------------------
    PermissionDef("leave.view_all", "Leave", "View all",
                  "See every employee's leave requests and their status."),
    PermissionDef("leave.view_own", "Leave", "View own",
                  "See only their own leave requests."),
    PermissionDef("leave.apply", "Leave", "Apply",
                  "Request leave for themselves."),
    PermissionDef("leave.approve", "Leave", "Approve",
                  "Approve or reject somebody else's leave. Nobody can approve their own, "
                  "whatever permissions they hold."),
    PermissionDef("holidays.manage", "Leave", "Manage holidays",
                  "Add and remove company holidays. A holiday stops a company-wide "
                  "zero-punch day from being reported as everybody being absent."),

    # --- Corrections -------------------------------------------------------
    PermissionDef("corrections.request", "Attendance", "Request a correction",
                  "Ask for a mistaken or missing punch to be fixed. The original "
                  "record is never altered — a correction is layered on top."),
    PermissionDef("corrections.view_all", "Attendance", "View all corrections",
                  "See correction requests from everyone."),
    PermissionDef("corrections.approve", "Attendance", "Approve corrections",
                  "Approve or reject a correction. Nobody can approve their own."),

    PermissionDef("settings.manage", "Administration", "Manage settings",
                  "Change shift times, the grace period, the minimum hours in a day, "
                  "and create departments."),
    PermissionDef("audit.view", "Administration", "View audit log",
                  "See who changed what and what it looked like before. Read-only — "
                  "the log itself cannot be edited or deleted through the system."),

    # --- Kiosk (Phase 3) -----------------------------------------------------
    PermissionDef("devices.manage", "Kiosk", "Manage kiosk devices",
                  "Register a new kiosk tablet, see the one-time device code, and revoke a "
                  "device that has left the building."),
    PermissionDef("face.enrol", "Kiosk", "Enrol face templates",
                  "Upload an employee's enrolment photo for the kiosk to recognise them. "
                  "Separate from employees.edit because it is biometric data with its own "
                  "consent requirement, not a profile detail."),
    PermissionDef("pin.generate", "Kiosk", "Generate kiosk PIN",
                  "Generate or regenerate an employee's kiosk PIN. Shown once, never stored "
                  "or logged in plain."),
]

ALL_CODES: set[str] = {p.code for p in PERMISSIONS}

# Seeded roles. Everything else is created through the API.
SEED_ROLES: dict[str, dict] = {
    "Owner": {
        "description": "Full access to everything, including roles and user accounts",
        "is_system": True,
        "permissions": sorted(ALL_CODES),
    },
    "Manager": {
        "description": "Sees the whole team and assigns work, but cannot manage access",
        "is_system": False,
        "permissions": [
            "employees.view_all", "employees.view_own",
            "attendance.view_all", "attendance.view_own",
            "tasks.view_all", "tasks.view_own", "tasks.assign",
            "tasks.create_own", "tasks.edit", "tasks.complete_own",
            "users.change_role",
            "leave.view_all", "leave.view_own", "leave.apply", "leave.approve",
            "corrections.request", "corrections.view_all", "corrections.approve",
            "face.enrol", "pin.generate",
        ],
    },
    "Employee": {
        "description": "Own record, own attendance, own tasks",
        "is_system": True,
        "permissions": [
            "employees.view_own",
            "attendance.view_own",
            "tasks.view_own", "tasks.create_own", "tasks.complete_own",
            "leave.view_own", "leave.apply",
            "corrections.request",
        ],
    },
}

File: attendance-admin/src/contracts/permissions.ts
/**
 * src/contracts/permissions.ts
 *
 * WHY THIS FILE EXISTS
 * --------------------
 * Permission codes are strings. In plain JavaScript a typo like
 * 'attendance.viewall' (missing underscore) compiles fine, returns false at
 * runtime, and silently shows the wrong UI to the wrong person. You find that
 * bug months later, in production.
 *
 * Declaring every valid code as a union type turns that typo into a red
 * squiggle in your editor, before you save the file.
 *
 * RULE: never write a permission string inline anywhere else. Import from here.
 */

export const PERMISSIONS = [
  // --- Employees ---------------------------------------------------------
  'employees.view_all',
  'employees.view_own',
  'employees.create',
  'employees.edit',

  // --- Attendance --------------------------------------------------------
  'attendance.view_all',
  'attendance.view_own',
  'attendance.punch_manual',

  // --- Tasks -------------------------------------------------------------
  'tasks.view_all',
  'tasks.view_own',
  'tasks.assign',
  'tasks.create_own',
  'tasks.edit',
  'tasks.complete_own',

  // --- Administration ----------------------------------------------------
  'roles.manage', // create roles, edit their permission sets
  'users.manage', // create logins, reset passwords, disable accounts
  'users.change_role', // move someone to a different role
  'leave.view_all',
  'leave.view_own',
  'leave.apply',
  'leave.approve',
  'holidays.manage',

  'corrections.request',
  'corrections.view_all',
  'corrections.approve',

  'settings.manage',
  'audit.view',

  // --- Kiosk (Phase 3) -----------------------------------------------------
  'devices.manage',
  'face.enrol',
  'pin.generate',
] as const;

/**
 * `as const` above freezes the array into readonly string literals. Without it
 * TypeScript widens the type to plain string[] and the entire safety benefit
 * disappears.
 *
 * This line turns the array into a union type:
 *   'employees.view_all' | 'employees.view_own' | ...
 */
export type Permission = (typeof PERMISSIONS)[number];

/**
 * WHY view_all AND view_own ARE SEPARATE PERMISSIONS
 * ---------------------------------------------------
 * The owner sees everyone's attendance; an employee sees only their own. With
 * a single 'attendance.view' permission you could not express that difference
 * and would end up hard-coding role names in the query layer — which breaks
 * the moment a custom role exists.
 *
 * With the split the scope comes from the permission alone:
 *   view_all -> no filter
 *   view_own -> filter to the logged-in person
 *
 * A 'view_team' variant slots in later without touching anything already
 * written.
 */

/** Grouping used to render the roles matrix screen. */
export const PERMISSION_MODULES = {
  Employees: PERMISSIONS.filter((p) => p.startsWith('employees.')),
  Attendance: PERMISSIONS.filter((p) => p.startsWith('attendance.')),
  Tasks: PERMISSIONS.filter((p) => p.startsWith('tasks.')),
  Administration: PERMISSIONS.filter(
    (p) =>
      p.startsWith('roles.') ||
      p.startsWith('users.') ||
      p.startsWith('settings.'),
  ),
  Kiosk: PERMISSIONS.filter(
    (p) => p.startsWith('devices.') || p.startsWith('face.') || p.startsWith('pin.'),
  ),
} satisfies Record<string, readonly Permission[]>;

/** Short column labels for the matrix header. */
export const PERMISSION_LABELS: Record<Permission, string> = {
  'employees.view_all': 'View all',
  'employees.view_own': 'View own',
  'employees.create': 'Create',
  'employees.edit': 'Edit',

  'attendance.view_all': 'View all',
  'attendance.view_own': 'View own',
  'attendance.punch_manual': 'Manual punch',

  'tasks.view_all': 'View all',
  'tasks.view_own': 'View own',
  'tasks.assign': 'Assign',
  'tasks.create_own': 'Add own',
  'tasks.edit': 'Edit',
  'tasks.complete_own': 'Complete own',

  'roles.manage': 'Manage roles',
  'users.manage': 'Manage users',
  'users.change_role': 'Change role',
  'leave.view_all': 'View all',
  'leave.view_own': 'View own',
  'leave.apply': 'Apply',
  'leave.approve': 'Approve',
  'holidays.manage': 'Manage holidays',
  'corrections.request': 'Request a correction',
  'corrections.view_all': 'View all corrections',
  'corrections.approve': 'Approve corrections',
  'settings.manage': 'Manage settings',
  'audit.view': 'View audit log',

  'devices.manage': 'Manage kiosk devices',
  'face.enrol': 'Enrol face templates',
  'pin.generate': 'Generate kiosk PIN',
};

/**
 * Plain-English meaning of every permission, shown in the roles matrix so an
 * owner is never guessing what a checkbox does.
 *
 * WHY PERMISSIONS CANNOT BE CREATED FROM THE UI
 * ----------------------------------------------
 * A permission is not data — it is a gate written into the code, in two places:
 * the screen that hides a control, and the API endpoint that refuses the
 * request. Inventing a new code here would produce a checkbox that no code
 * consults: it would appear to lock something down while locking down nothing,
 * which is worse than having no checkbox at all.
 *
 * New permissions arrive with new features, because the feature is what
 * enforces them. What IS freely creatable is a ROLE — any combination of the
 * permissions below, named however you like.
 */
export const PERMISSION_DESCRIPTIONS: Record<Permission, string> = {
  'employees.view_all':
    'See the whole company directory, with everyone\u2019s contact details, department and login status.',
  'employees.view_own':
    'See only their own employee record. Everyone needs this to open their profile.',
  'employees.create': 'Add a new person to the directory.',
  'employees.edit':
    'Change anyone\u2019s details \u2014 name, contact, department, position, or mark them inactive.',

  'attendance.view_all':
    'See every employee\u2019s check-in and check-out records, and the exception list. This is what turns the dashboard into the company-wide view.',
  'attendance.view_own':
    'See only their own attendance history. Everyone needs this to check their own hours.',
  'attendance.punch_manual':
    'Record a check-in or check-out by hand, for when the kiosk misses someone. Entries are marked as manual and stay visible as such.',

  'tasks.view_all': 'See tasks assigned to everyone, and filter the board by person.',
  'tasks.view_own': 'See only tasks assigned to themselves.',
  'tasks.assign': 'Create a task for somebody else and set its due date and priority.',
  'tasks.create_own':
    'Add a task to their own list. Useful when people plan their own day rather than waiting to be assigned work.',
  'tasks.edit': 'Change a task\u2019s title, description, due date or priority after it was created.',
  'tasks.complete_own': 'Tick off a task assigned to them.',

  'roles.manage':
    'Create roles and change what every role can do \u2014 including this screen. At least one role in use must always keep it.',
  'users.manage':
    'Create login accounts, reset passwords, and enable or disable someone\u2019s access.',
  'users.change_role':
    'Move someone to a different role. You can only assign roles that are no more powerful than your own, and you cannot change your own role \u2014 otherwise anyone with this could promote themselves.',
  'settings.manage':
    'Change company policy: shift times, the grace period before someone counts as late, the minimum hours in a day, and create departments.',
  'leave.view_all': "See every employee's leave requests and their status.",
  'leave.view_own': 'See only their own leave requests.',
  'leave.apply': 'Request leave for themselves.',
  'leave.approve':
    'Approve or reject somebody else\u2019s leave. Nobody can approve their own, whatever permissions they hold.',
  'holidays.manage':
    'Add and remove company holidays. A holiday stops a company-wide zero-punch day from being reported as everybody being absent.',
  'corrections.request':
    'Ask for a mistaken or missing punch to be fixed. The original record is never altered \u2014 a correction is layered on top.',
  'corrections.view_all': 'See correction requests from everyone.',
  'corrections.approve':
    'Approve or reject a correction. Nobody can approve a correction to their own attendance.',
  'audit.view':
    'See who changed what and what it looked like before. Read-only \u2014 the log cannot be edited or deleted through the system, which is what makes it evidence.',

  'devices.manage':
    'Register a new kiosk tablet, see the one-time device code, and revoke a device that has left the building.',
  'face.enrol':
    'Upload an employee\u2019s enrolment photo for the kiosk to recognise them. Separate from employees.edit because it is biometric data with its own consent requirement, not a profile detail.',
  'pin.generate':
    'Generate or regenerate an employee\u2019s kiosk PIN. Shown once, never stored or logged in plain.',
};